src/main/java/org/opensearch/geospatial/index/query/xyshape/XYShapeQueryVisitor.java

```diff
@@ -0,0 +1,121 @@
+
+import static org.opensearch.geospatial.index.common.xyshape.XYShapeConverter.toXYCircle;
+import static org.opensearch.geospatial.index.common.xyshape.XYShapeConverter.toXYLine;
+import static org.opensearch.geospatial.index.common.xyshape.XYShapeConverter.toXYPoint;
+import static org.opensearch.geospatial.index.common.xyshape.XYShapeConverter.toXYPolygon;
+import static org.opensearch.geospatial.index.common.xyshape.XYShapeConverter.toXYRectangle;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.lucene.geo.XYGeometry;
+import org.opensearch.geometry.Circle;
+import org.opensearch.geometry.Geometry;
+import org.opensearch.geometry.GeometryCollection;
+import org.opensearch.geometry.GeometryVisitor;
+import org.opensearch.geometry.Line;
+import org.opensearch.geometry.LinearRing;
+import org.opensearch.geometry.MultiLine;
+import org.opensearch.geometry.MultiPoint;
+import org.opensearch.geometry.MultiPolygon;
+import org.opensearch.geometry.Point;
+import org.opensearch.geometry.Polygon;
+import org.opensearch.geometry.Rectangle;
+import org.opensearch.geometry.ShapeType;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.index.query.QueryShardException;
+
+/**
+ * Geomtery Visitor to create a query to find all cartesian XYShapes
+ * that comply ShapeRelation with all other XYShapes objects
+ */
+public class XYShapeQueryVisitor implements GeometryVisitor<List<XYGeometry>, RuntimeException> {
+
+    private final String name;
+    private final QueryShardContext context;
+
+    public XYShapeQueryVisitor(String name, QueryShardContext context) {
+        this.name = name;
+        this.context = context;
+    }
+
+    @Override
+    public List<XYGeometry> visit(Circle circle) throws RuntimeException {
+        Objects.requireNonNull(circle, ShapeType.CIRCLE + " cannot be null");
+        // XYShape don't support indexing Circle, but, can perform query to identify list of shapes
+        // that interact with a provided circle
+        return List.of(toXYCircle(circle));
+    }
+
+    @Override
+    public List<XYGeometry> visit(GeometryCollection<?> geometryCollection) throws RuntimeException {
+        Objects.requireNonNull(geometryCollection, " Geometry collection cannot be null");
+        return visitCollection(geometryCollection);
+    }
+
+    @Override
+    public List<XYGeometry> visit(Line line) throws RuntimeException {
+        Objects.requireNonNull(line, "Line cannot be null");
+        return List.of(toXYLine(line));
+    }
+
+    @Override
+    public List<XYGeometry> visit(LinearRing linearRing) throws RuntimeException {
+        throw new QueryShardException(this.context, "Field [" + this.name + "] found an unsupported shape LinearRing");
+    }
+
+    @Override
+    public List<XYGeometry> visit(MultiLine multiLine) throws RuntimeException {
+        Objects.requireNonNull(multiLine, "Multi line cannot be null");
```

Comment:
nit: should we change "line" to "Line" to have it consistent with other methods

Reply:
Ack